packages/compiler-cli/src/perform_compile.ts

```
@@ -118,9 +118,133 @@ export function calcProjectFileAndBasePath(project: string):
+    renderer2BackPatching: false,
+    generateCodeForLibraries: false,
+    skipMetadataEmit: false,
+    skipTemplateCodegen: true,
+    enableLegacyTemplate: false,
+    preserveWhitespaces: false,
+    fullTemplateTypeCheck: true,
+  },
+  package: {
+    generateRenderer2Factories: false,
+    renderer2BackPatching: false,
+    generateCodeForLibraries: false,
+    enableIvy: false,
+    skipMetadataEmit: false,
+    skipTemplateCodegen: true,
+    enableLegacyTemplate: false,
+    preserveWhitespaces: false,
+    fullTemplateTypeCheck: true,
+  }
+};
+
+// The values these options must be if they "target" is used.
+const enforced: TargetOptions = {
+  application: {
+    generateRenderer2Factories: true,
+    renderer2BackPatching: true,
+  },
+  library: {
+    generateRenderer2Factories: false,
+    renderer2BackPatching: false,
+    generateCodeForLibraries: false,
+    skipMetadataEmit: false,
+    skipTemplateCodegen: true,
+  },
+  package: {
+    generateRenderer2Factories: false,
+    renderer2BackPatching: false,
+    generateCodeForLibraries: false,
+    enableIvy: false,
+    skipMetadataEmit: false,
+    skipTemplateCodegen: true,
+
+    // `undefined` means that it must be set but no value set value is enforced.
+    flatModuleId: undefined,
+    flatModuleOutFile: undefined,
+  }
+};
+
+export interface NgOptions {
+  options: api.CompilerOptions;
+  errors: Diagnostics;
+}
+
+export function processTargetAndValidateNgCompilerOptions(options: any): NgOptions {
+  const errors: (api.Diagnostic | ts.Diagnostic)[] = [];
+  function reportError(messageText: string) {
+    errors.push({
+      category: ts.DiagnosticCategory.Error,
+      messageText: `Angular compiler configuration error, ${messageText}`,
+      source: api.SOURCE,
+      code: api.UNKNOWN_ERROR_CODE
+    });
+  }
+
+  // If generateRenderer2Factories is specified then renderer2BackPatching defaults to `true`
+  if (options.generateRenderer2Factories) {
```

Does this silently override what the user specified? We should throw an error if the user specifically asks for `genreateRender2Factories` but sets `render2BackPatching` to `false`.

---

The changes the default as what they provide is supplied after in the next statement.

An error is reported below if they get this wrong.